SurprisesManager.cpp

```diff
@@ -0,0 +1,322 @@
+#include "SurprisesManager.h"
+
+using namespace SurprisesConst;
+using namespace WorldConst;
+
+SurprisesManager::SurprisesManager(World* world, const char* texture)
+{
+	if (!this->texture.loadFromFile(texture))
+	{
+		cout << "error loading surpriseTexture";
+		exit(-1);
+	}
+
+	srand(time(NULL));
+
+	this->world = world;
+
+	InitSprite(randomSurpriseSprite, SurpriseSprite::RANDOM);
+	InitSprite(bombSupplySurpriseSprite, SurpriseSprite::BOMB_SUPPLY);
+	InitSprite(blastRadiusSurpriseSprite, SurpriseSprite::BLAST_RADIUS);
+	InitSprite(speedSurpriseSprite, SurpriseSprite::SPEED);
+	InitSprite(invincibleSurpriseSprite, SurpriseSprite::INVINCIBLE);
+	//InitSprite();
+
+	//PrintSurpriseMap();
+
+	InitSurpriseMap();
+
+	GenerateSurprises();
+}
+
+void SurprisesManager::InitSprite(sf::Sprite& sprite, MatPos pos)
+{
+	sprite.setTexture(this->texture);
+	sprite.setTextureRect(
+		sf::IntRect(
+			pos.c * SurpriseSprite::FRAME_WIDTH,
+			pos.l * SurpriseSprite::FRAME_HEIGHT,
+			SurpriseSprite::FRAME_WIDTH,
+			SurpriseSprite::FRAME_HEIGHT
+		)
+	);
+	sprite.setScale(
+		CELL_WIDTH / SurpriseSprite::FRAME_WIDTH,
+		CELL_HEIGHT / SurpriseSprite::FRAME_HEIGHT
+	);
+}
+
+void SurprisesManager::InitSprite()
+{
+	randomSurpriseSprite.setTexture(this->texture);
+	randomSurpriseSprite.setTextureRect(
+		sf::IntRect(
+			SurpriseSprite::RANDOM.c * SurpriseSprite::FRAME_WIDTH,
+			SurpriseSprite::RANDOM.l * SurpriseSprite::FRAME_HEIGHT,
+			SurpriseSprite::FRAME_WIDTH,
+			SurpriseSprite::FRAME_HEIGHT
+		)
+	);
+	randomSurpriseSprite.setScale(
+		CELL_WIDTH / SurpriseSprite::FRAME_WIDTH,
+		CELL_HEIGHT / SurpriseSprite::FRAME_HEIGHT
+	);
+}
+
+void SurprisesManager::InitSurpriseMap()
+{
+	for (int i = 0; i < NL; i++)
+	{
+		for (int j = 0; j < NC; j++)
+		{
+			surpriseMap[i][j] = SurpriseType::NONE;
+		}
+	}
+}
+
+void SurprisesManager::PrintSurpriseMap()
+{
+	for (int i = 0; i < NL; i++)
+	{
+		for (int j = 0; j < NC; j++)
+		{
+			cout << surpriseMap[i][j] << " ";
+		}
+		cout << endl;
+	}
+}
+
+void SurprisesManager::GenerateInvincibleSurprises()
+{
+	int l, c;
+
+	for (int i = (4 * SURPRISE_COUNT) / 5; i < SURPRISE_COUNT; i++)
+	{
+		l = rand() % (NL);
+		c = rand() % (NC);
+
+		while (!world->IsCellBox(l, c) || surpriseMap[l][c] != SurpriseType::NONE)
+		{
+			l = rand() % (NL);
+			c = rand() % (NC);
+		}
+
+		surpriseMap[l][c] = SurpriseType::INVINCIBLE;
+	}
+}
+
+void SurprisesManager::GenerateSpeedSurprises()
+{
+	int l, c;
+
+	for (int i = (3 * SURPRISE_COUNT) / 5; i < (4 * SURPRISE_COUNT) / 5; i++)
+	{
+		l = rand() % (NL);
+		c = rand() % (NC);
+
+		while (!world->IsCellBox(l, c) || surpriseMap[l][c] != SurpriseType::NONE)
+		{
+			l = rand() % (NL);
+			c = rand() % (NC);
+		}
+
+		surpriseMap[l][c] = SurpriseType::SPEED;
+	}
+}
+
+void SurprisesManager::GenerateBlastRadiusSurprises()
```

Comment:
Too many methods that do the same thing.
Just make one, that takes how many surprises need to be generated (based on the surprise's appearance probability) and the surprise type